File: Programmers_make_all_zeros.js
function isValid(a) {
    let sum = a.reduce((stack, val) => stack+val, 0);
    return sum === 0 ? true : false;
}

function getDicSize(target, n) {
    for(let i = 0; i < n.length; ++i) target[target.length] = [];    
    return target;
}

function setConnection(dic, edges) {
    for(let i = 0; i < edges.length; i++) {
        const [x1, x2] = edges[i];
        
        dic[x1].push(x2);
        dic[x2].push(x1);
    }
}

function getVisit(visit, a) {
    for(let i = 0; i < a.length; ++i) visit[visit.length] = false;
}

function dfs(dic, visit, a) {
    let stack = [[0,-1]];
    let ret = 0n;
    
    while(stack.length) {
        const [node, parent] = stack.pop();
        
        if(visit[node]) {
            ret += BigInt(Math.abs(a[node]));
            a[parent] += a[node];
            a[node] = 0;
            
            continue;
        }
        
        visit[node] = true;
        stack.push([node, parent]);
        
        for(let i = 0; i < dic[node].length; ++i) {
            if(visit[dic[node][i]]) continue;
            
            stack.push([dic[node][i], node]);
        }
    }
    
    return ret;
}

function solution(a, edges) {
    let dic = [];
    let visit = [];
    
    if(!isValid(a)) return -1;
    getDicSize(dic, a);
    getVisit(visit, a);
    setConnection(dic, edges);
    
    let answer = dfs(dic, visit, a)
    
    return answer;
}

File: Programmers_make_all_zeros.cpp
#include <bits/stdc++.h>

using namespace std;

void dfs(long long now, long long pre, vector<long long> &a, vector<long long> edges[], long long &answer) {
    
    for(long long i = 0; i < edges[now].size(); ++i) {
        if(pre == edges[now][i]) continue;
        
        dfs(edges[now][i], now, a, edges, answer);
    }
    
    answer += abs(a[now]);
    
    if(pre == -1) return;
    a[pre] += a[now];
}

void setEdges(vector<long long> edges[], vector<vector<int>> info) {
    for(long long i = 0; i < info.size(); ++i) {
        long long x1 = info[i][0];
        long long x2 = info[i][1];
        
        edges[x1].push_back(x2);
        edges[x2].push_back(x1);
    }
}

long long solution(vector<int> a, vector<vector<int>> edges) {
    long long answer = 0;
    
    vector<long long> b(a.begin(), a.end());
    if(accumulate(b.begin(), b.end(), 0ll) != 0) return -1;
    
    vector<long long> t_edges[300000];
    setEdges(t_edges, edges);
    
    dfs(0, -1, b, t_edges, answer);
    
    return answer;
}